x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java

```diff
@@ -392,41 +392,31 @@ public static Query doTranslate(In in, TranslatorHandler handler) {
         }
 
         private static Query translate(In in, TranslatorHandler handler) {
-            Query q;
-            if (in.value() instanceof FieldAttribute) {
-                // equality should always be against an exact match (which is important for strings)
-                FieldAttribute fa = (FieldAttribute) in.value();
-                DataType dt = fa.dataType();
-
-                List<Expression> list = in.list();
-                Set<Object> set = new LinkedHashSet<>(CollectionUtils.mapSize(list.size()));
-                list.forEach(e -> {
-                    // TODO: this needs to be handled inside the optimizer
-                    if (DataTypes.isNull(e.dataType()) == false) {
-                        set.add(handler.convert(valueOf(e), dt));
-                    }
-                });
-
-                if (DataTypes.isDateTime(dt)) {
-                    DateFormatter formatter = DateFormatter.forPattern(DATE_FORMAT);
-
-                    q = null;
-                    for (Object o : set) {
-                        assert o instanceof ZonedDateTime : "expected a ZonedDateTime, but got: " + o.getClass().getName();
-                        // see comment in Ranges#doTranslate() as to why formatting as String is required
-                        String zdt = formatter.format((ZonedDateTime) o);
-                        RangeQuery right = new RangeQuery(
-                            in.source(), fa.exactAttribute().name(),
-                            zdt, true, zdt, true, formatter.pattern(), in.zoneId());
-                        q = q == null ? right : new BoolQuery(in.source(), false, q, right);
+            assert in.value() instanceof FieldAttribute;
+
+            Set<Object> terms = new LinkedHashSet<>();
+            List<Query> queries = new ArrayList<>();
+
+            for (Expression rhs : in.list()) {
+                if (DataTypes.isNull(rhs.dataType()) == false) {
+                    // delegates to BinaryComparisons translator to ensure consistent handling of date and time values
+                    Query query = BinaryComparisons.translate(new Equals(in.source(), in.value(), rhs, in.zoneId()), handler);
+
+                    if (query instanceof TermQuery) {
+                        terms.add(((TermQuery) query).value());
+                    } else {
+                        queries.add(query);
                     }
-                } else {
-                    q = new TermsQuery(in.source(), fa.exactAttribute().name(), set);
                 }
-            } else {
-                q = new ScriptQuery(in.source(), in.asScript());
             }
-            return q;
+
+            String fieldName = ((FieldAttribute) in.value()).exactAttribute().name();
+            if (terms.isEmpty() == false) {
+                queries.add(new TermsQuery(in.source(), fieldName, terms));
+            }
+
+            return queries.stream()
```

Comment:
I don't recall creating a `BoolQuery` by hand anywhere else - usually this is done in the optimizer not the translator which just picks it up.
Here it is being used depending on the resulting query - however that will be the case only in case of datetime field which we know in advanced.

That is my preference is to handle this breaking / rewriting of the query at a higher-level, in the plan as oppose to during the translation level which should be simple and not worry about combining things.

Reply:
There were some concerns I had with doing this in the optimizer:
- The fact that some field/value combination cannot benefit from the `IN` optimization seemed very execution specific to me. Guarding against this in the optimizer would mean that some execution specific knowledge (wether a comparison is executed as a term query) leaks into the optimizer.
- This by itself is not enough to make `a IN (...)` and `a = ... OR a = ...` behave consistently. To consistently support date comparisons in `IN`, the optimizer would need a rule to split up `IN` with comparisons that are not executed as term queries. This is not really an optimization though.
- To me it looks somewhat similar to what we're doing in `QueryFolder.FoldFilter`. There, we also combine expressions from multiple `Filter` plans.

I'm replacing the manual creation of the `BoolQuery` though, as we already have specialized methods in `ExpressionTranslators` for this.